table: make update_effective_replication_map sync again

bhalevy opened the pull request:
Commit f2ff701489995fdc84ea7ca0addbec2918c35be4 introduced
a yield in update_effective_replication_map that might
cause the storage_group manager to be inconsistent with the
new effective_replication_map (e.g. if yielding right
before calling `handle_tablet_split_completion`.

Also, yielding inside storage_service::replicate_to_all_cores
update loop means that base tables and their views
aren't updated atomically, that caused scylladb/scylladb#17786

This change essentially reverts f2ff701489995fdc84ea7ca0addbec2918c35be4
and makes handle_tablet_split_completion synchronous too.
The stopped compaction groups future is kept as a member and
storage_group_manager::stop() consumes this future during table::stop().

- storage_service: replicate_to_all_cores: update base and view tables atomically
    
Currently, the loop updating all tables (including views) with the
new effective_replication_map may yield, and therefore expose
a state where the base and view tables effective_replication_map
and topology are out of sync (as seen in scylladb/scylladb#17786)

To prevent that, loop over all base tables and for each table
update the base table and all views atomically, without yielding,
and so allow yielding only between base tables.

* Regression was introduced in f2ff701489995fdc84ea7ca0addbec2918c35be4, so backport is required to 6.x, 2024.2

raphaelsc commented:
"The table::_erm is update first and only then its storage_group manager is updated asynchronously, and since the latter might yield, it may expose a partially updated state of the storage_groups." not the case today, but I agree with the fragility. today, we make sure replica state always reflect the erm state. I had some concerns with futurizing update_effective_replication_map() due to that.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13603/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/508/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5147/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13164/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2039/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13603/)
#### Failed Tests (38/35969):
- [test_concurrent_tablet_repair_and_split.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_concurrent_tablet_repair_and_split_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_tablet_repair_and_split.debug.4)
- [test_concurrent_tablet_repair_and_split.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_concurrent_tablet_repair_and_split_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_tablet_repair_and_split.debug.5)
- [test_concurrent_tablet_repair_and_split.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_concurrent_tablet_repair_and_split_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_tablet_repair_and_split.debug.6)
- [test_concurrent_tablet_repair_and_split.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_concurrent_tablet_repair_and_split_dev_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_tablet_repair_and_split.dev.4)
- [test_concurrent_tablet_repair_and_split.dev.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_concurrent_tablet_repair_and_split_dev_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_tablet_repair_and_split.dev.5)
- [test_concurrent_tablet_repair_and_split.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_concurrent_tablet_repair_and_split_dev_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_concurrent_tablet_repair_and_split.dev.6)
- [test_correctness_of_tablet_split_finalization_after_restart.debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_correctness_of_tablet_split_finalization_after_restart_debug_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_correctness_of_tablet_split_finalization_after_restart.debug.4)
- [test_correctness_of_tablet_split_finalization_after_restart.debug.5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_correctness_of_tablet_split_finalization_after_restart_debug_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_correctness_of_tablet_split_finalization_after_restart.debug.5)
- [test_correctness_of_tablet_split_finalization_after_restart.debug.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_correctness_of_tablet_split_finalization_after_restart_debug_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_correctness_of_tablet_split_finalization_after_restart.debug.6)
- [test_correctness_of_tablet_split_finalization_after_restart.dev.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13603/testReport/junit/topology_experimental_raft/test_tablets/Tests___Unit_Tests___test_correctness_of_tablet_split_finalization_after_restart_dev_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_correctness_of_tablet_split_finalization_after_restart.dev.4)
#### Build Details:
- Duration: 5 hr 28 min
- Builder: spider3.cloudius-systems.com

bhalevy commented:
In v2 (https://github.com/scylladb/scylladb/commit/f6e492b7af20e8458571dec3e0a75078e3f0cd4d):
- rebased on top of 7e2875d6489d2e6d4e22dce721f634bab79847bf
- Simplified, table: make update_effective_replication_map sync again
- storage_service: replicate_to_all_cores: update base and view tables atomically
  - kept, in a simplified form, with sync call to update_effective_replication_map
- storage_servive: replicate_to_all_cores: clear_gently pending erms
  - tiny improvement for rollback on error path

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13661/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/550/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5212/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13228/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2104/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13661/)
#### Failed Tests (2/36080):
- [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13661/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)
- [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13661/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)
#### Build Details:
- Duration: 3 hr 57 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13661/) ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/550/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5212/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13228/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2104/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13661/)
> 
> #### Failed Tests (2/36080):
> * [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13661/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)
> * [test_simple_removenode_5](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13661/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_removenode_5_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_removenode_5)

Unrelated: https://github.com/scylladb/scylladb/issues/20814

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13664/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/553/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2108/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5216/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13232/)
#### Build Details:
- Duration: 1 hr 6 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
@wmitros ping, please review

wmitros reviewed (approved):
Looks good to me.
I'm not completely sure if removing the yields could cause stalls, but there shouldn't be that many tables in a keyspace and I don't think we had stalls here before https://github.com/scylladb/scylladb/commit/f2ff701489995fdc84ea7ca0addbec2918c35be4 so that's probably fine.
I added a suggestion that might be worth considering, but the patch should work even without it

bhalevy commented on service/storage_service.cc:
```diff
@@ -3157,9 +3164,20 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
             for (auto it = table_erms.begin(); it != table_erms.end(); ) {
                 auto& cf = db.find_column_family(it->first);
                 cf.update_effective_replication_map(std::move(it->second));
+                for (const auto& view_ptr : cf.views()) {
+                    const auto& view_id = view_ptr->id();
+                    auto view_it = view_erms.find(view_id); 
+                    if (view_it == view_erms.end()) {
+                        throw std::runtime_error(format("Could not find pending effective_replication_map for view {}.{} id={}", view_ptr->ks_name(), view_ptr->cf_name(), view_id));
+                    }
+                    auto& view = db.find_column_family(view_id);
+                    view.update_effective_replication_map(std::move(view_it->second));
+                    view_erms.erase(view_it);
+                }
```
You're right.
Now that we prevent this yield on purpose I can remove that test scenario and instead add a deliberate maybe_yield call here to prevent stalls with many tables.

bhalevy replied:
Done in v3 ([2f1bb11](https://github.com/scylladb/scylladb/commit/2f1bb11548d8796c227cf0ad2e3e2304631ac6ba))

bhalevy commented:
In v3 ([2f1bb11](https://github.com/scylladb/scylladb/commit/2f1bb11548d8796c227cf0ad2e3e2304631ac6ba)):
- rebased
- test_mv_topology_change: drop delay_after_erm_update injection case

avikivity commented:
Are all commits part of the fix, or some are enhancements?

avikivity commented on service/storage_service.cc:
```diff
@@ -3049,7 +3049,9 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
     std::vector<std::unordered_map<sstring, locator::vnode_effective_replication_map_ptr>> pending_effective_replication_maps;
     pending_effective_replication_maps.resize(smp::count);
     std::vector<std::unordered_map<table_id, locator::effective_replication_map_ptr>> pending_table_erms;
```
> Currently, the loop updating all tables (including views) with the
> new effective_replication_map may yield, and therefore expose
> a state where the base and view tables effective_replication_map
> and topology are out of sync (as seen in https://github.com/scylladb/scylladb/issues/17786)
> 
> To prevent that, loop over all base tables and for each table
> update the base table and all views atomically, without yielding,
> and so allow yielding only between base tables.

This solution works, but it's fragile: we recognize one interdependency in the schema, and plug that hole.

Cannot we capture the entire schema atomically, then work on it in our leisure? (typically a database schema refers to all tables, not just one).

bhalevy replied:
> > Currently, the loop updating all tables (including views) with the
> > new effective_replication_map may yield, and therefore expose
> > a state where the base and view tables effective_replication_map
> > and topology are out of sync (as seen in #17786)
> > To prevent that, loop over all base tables and for each table
> > update the base table and all views atomically, without yielding,
> > and so allow yielding only between base tables.
> 
> This solution works, but it's fragile: we recognize one interdependency in the schema, and plug that hole.
> 
> Cannot we capture the entire schema atomically, then work on it in our leisure? (typically a database schema refers to all tables, not just one).

It's possible using a prepare phase for calculating the modified state and then applying it atomically to all tables, but it is more complicated and higher risk, and so less backportable.
So I suggest starting with this simple approach and following up with a comprehensive fix going forward.

avikivity commented on service/storage_service.cc:
```diff
@@ -3118,7 +3120,11 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
+                    pending_view_erms[this_shard_id()].emplace(id, std::move(erm));
+                } else {
+                    pending_table_erms[this_shard_id()].emplace(id, std::move(erm));
+                }
```
Is this loop not a huge stall with large table counts (independently of the patch)? Or with a single table with large numbers of tablets?

bhalevy replied:
Good question, we haven't see an stalls in our tests, but maybe we didn't test with a large number of tablet-based tables.
`make_replication_map` can take significant amount of cpu (like `create_effective_replication_map`), and we should allow yielding in the loop as well.

bhalevy replied:
Sorry, after looking deeper into the code, I realized that `make_replication_map` just allocates and constructs a `tablet_effective_replication_map` using the shared replication_strategy_ptr and token_metadata_ptr so it is not cpu intensive, though allocating thousands of objects via `tablet_aware_replication_strategy::do_make_replication_map` might stall as well.

bhalevy replied:
https://github.com/scylladb/scylladb/pull/21924

avikivity commented on service/storage_service.cc:
```diff
@@ -3178,6 +3197,10 @@ future<> storage_service::replicate_to_all_cores(mutable_token_metadata_ptr tmpt
             }
 
+            if (!view_erms.empty()) {
+                throw std::runtime_error(fmt::format("Found orphaned pending effective_replication_maps for the following views: {}", boost::adaptors::keys(view_erms)));
```
Please don't add more boost stuffs

bhalevy replied:
Fixed in v4 ([cf0d00a](https://github.com/scylladb/scylladb/commit/cf0d00ac904a20fb770ff85a4e7da11db99130b2)):

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13773/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13773/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_topology_change
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5339/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13348/)
:white_check_mark: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/44/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/642/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2224/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13773/)
#### Build Details:
- Duration: 4 hr 47 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
In v4 ([cf0d00a](https://github.com/scylladb/scylladb/commit/cf0d00ac904a20fb770ff85a4e7da11db99130b2)):
- rebased
- used `std::views::keys` rather than `boost::adaptors::keys`.

bhalevy commented:
> Are all commits part of the fix, or some are enhancements?

"storage_service: replicate_to_all_cores: clear_gently pending erms" adds a missing (pre-existing) call to `clear_gently` along with a new one that is part of this fix.
Otherwise all patches are part of the fix.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13833/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13833/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_mv_topology_change
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/697/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5420/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13431/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2302/)
:x: - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/99/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13833/)
#### Build Details:
- Duration: 4 hr 42 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13833/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13833/) **The following new/updated tests ran 100 times for each mode:** 🔹 topology_custom/test_mv_topology_change ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/697/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5420/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13431/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2302/) ❌ - [Offline-installer Artifact Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-offline-installer/99/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13833/)

Unrelated: https://github.com/scylladb/scylla-pkg/issues/4690

bhalevy commented:
@avikivity please re-review and consider queuing

bhalevy commented:
@avikivity ping, please re-review

bhalevy commented:
> @avikivity ping, please re-review

@avikivity ping^
